plugins/org.eclipse.xtext.generator/src/org/eclipse/xtext/generator/parser/antlr/AntlrGrammarComparator.xtend

```diff
@@ -0,0 +1,165 @@
+	// as well as character sequences containing no white space (\s), no single quotes ('),
+	//  no double quotes ("), and none of SINGLE_CHAR_TOKENS 
+	private static val TOKEN = '''�SINGLE_CHAR_TOKENS.join('|')�|�QUOTED_TOKENS�|[^\s'"�SINGLE_CHAR_TOKENS.join�]+''';
+	
+	// platform neutral definition of newline tokens, they are distinguished for counting lines
+	private static val NEWLINE = "\\r?\\n|\\r"
+	
+	// other white space to be ignored, is distinguished for testing for any unmatched character sequences
+	private static val WS = "( |\\t)+"		
+	
+	private val p_token = Pattern.compile(TOKEN)
+	private val p_newline = Pattern.compile(NEWLINE)
+	private val p_ws = Pattern.compile(WS)
+	private val compoundPattern = Pattern.compile('''(�TOKEN�)|(�NEWLINE�)|(�WS�)''', Pattern.MULTILINE) 		
+
+	
+	/**
+	 * Performs the actual comparison of given and expected grammar.
+	 * 
+	 * @return {@link Pair} containing the number of lines of the tested grammar (key)
+	 * 			and the referenced grammar (value) for logging purposes
+	 */
+	public def compareGrammars(CharSequence grammar, CharSequence grammarReference,
+			AbstractErrorHandler errorHandler) {
+
+		val compoundMatcher = compoundPattern.matcher(grammar)
+		val compoundMatcherReference = compoundPattern.matcher(grammarReference)
+		
+		// these values are used to check whether any character sequences have not been matched
+		var previousEnd = 0;
+		var previousEndReference = 0;
+		
+		var continue = true
+		var continueReference = true 
+		
+		while (continue || continueReference) {
+			if (continue) {
+				errorHandler.treatingReferenceGrammar = false
+				
+				val res = compoundMatcher.nextToken(previousEnd, errorHandler)
+				continue = res.key
+				errorHandler.lineNumber += res.value
+				
+				if (continue) {
+					previousEnd = compoundMatcher.end
+				}
+			}
+			val match = if (continue) compoundMatcher.group else "��eof��"
+			
+			if (continueReference) {
+				errorHandler.treatingReferenceGrammar = true
+				
+				val res = compoundMatcherReference.nextToken(previousEndReference, errorHandler)
+				continueReference = res.key
+				errorHandler.lineNumberReference += res.value
+				
+				if (continueReference) {
+					previousEndReference = compoundMatcherReference.end
+				}
+			}			 
+			
+			val matchReference = if (continueReference) compoundMatcherReference.group else "��eof��"
+			
+			if (matchReference != match) {
+				errorHandler.handleMismatch(match, matchReference)
+			}
+		}
+	}
+	
+	/**
+	 * continue matching until the next token is found, count newlines while doing so.
+	 * 
+	 * @return the number of newlines passed while searching 
+	 */
+	def private nextToken(Matcher matcher, int previousEnd, AbstractErrorHandler errorHandler) {
+		var newlineCounter = 0;
+		var thePreviousEnd = previousEnd
+		
+		while (matcher.find()) {
+			if (matcher.start() != thePreviousEnd) {
+				errorHandler.handleUnexpectedCharSequence(newlineCounter)
```

Comment:
If I understand correctly, this indicates a programming error in the Comparator, right? Then it should just throw an AssertionError instead of calling the ErrorHandler.